Rename labels

**Motivation**

Being able to rename labels would make it easier to work on long documents without resorting to search and replace to change figures/equations labels.

**Description**

Implement LSP rename capability for labels.

---
Hello! Not sure if this is intentional or a bug, on my side at least, the rename command does not work for labels.

Hello! I just tested on the main branch as well as the new extension from b67cec2. 
On my side I cannot rename a label either in Code or Helix rename a label.
```typ
#figure([Test]) <fig:test>
@fig:test
```
```
❯ tinymist --version
tinymist 
Build Timestamp:     2024-08-20T10:23:47.000000000Z
Build Git Describe:  v0.11.19-13-gb67cec2
Commit SHA:          b67cec2918db289bff1bf2c2dfe7d3e3206e819c
Commit Date:         None
Commit Branch:       None
Cargo Target Triple: x86_64-unknown-linux-gnu
Typst Version:       0.11.1
```

```
❯ hx --version
helix 24.7 (b90ec5c7)
```

```
Version: 1.93.0-insider
Commit: 922413f6d97e16c05b565398f33d95c306b1ceb7
Date: 2024-08-13T05:04:09.073Z
Electron: 30.3.1
ElectronBuildId: 9960165
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Linux x64 6.10.5-zen1-1-zen
```


Helix logs
```
2024-08-20T12:40:44.486 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:44Z INFO  sync_lsp] handling textDocument/prepareRename - (3) at Instant { tv_sec: 139358, tv_nsec: 262798420 }\n"
2024-08-20T12:40:44.486 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:44Z INFO  tinymist_query::prepare_rename] prepare_rename: not a function on function definition site: Some(FigureElem { body: Text(Test1), placement: Some(None), caption: Some(None), kind: Some(Custom(Elem(Element(image)))), supplement: Some(Custom(Some(Content(Text(Figure))))), numbering: Some(Some(Pattern(NumberingPattern { pieces: [(\"\", Arabic, Lower)], suffix: \"\", trimmed: false }))), gap: Some(0.65em), outlined: Some(true), counter: Some(Some(Counter(Selector(Elem(Element(figure), Some([(3, Func(image))])))))) })\n"
2024-08-20T12:40:44.486 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:44Z INFO  sync_lsp] handled  textDocument/prepareRename - (3) in 286.39µs\n"
2024-08-20T12:40:45.853 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:45Z INFO  sync_lsp] handling textDocument/prepareRename - (4) at Instant { tv_sec: 139359, tv_nsec: 629580277 }\n"
2024-08-20T12:40:45.853 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:45Z INFO  tinymist_query::prepare_rename] prepare_rename: not a function on function definition site: Some(FigureElem { body: Text(Test1), placement: Some(None), caption: Some(None), kind: Some(Custom(Elem(Element(image)))), supplement: Some(Custom(Some(Content(Text(Figure))))), numbering: Some(Some(Pattern(NumberingPattern { pieces: [(\"\", Arabic, Lower)], suffix: \"\", trimmed: false }))), gap: Some(0.65em), outlined: Some(true), counter: Some(Some(Counter(Selector(Elem(Element(figure), Some([(3, Func(image))])))))) })\n"
2024-08-20T12:40:45.853 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:45Z INFO  sync_lsp] handled  textDocument/prepareRename - (4) in 279.60µs\n"
2024-08-20T12:40:46.486 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:46Z INFO  sync_lsp] handling textDocument/prepareRename - (5) at Instant { tv_sec: 139360, tv_nsec: 262299484 }\n"
2024-08-20T12:40:46.486 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:46Z INFO  tinymist_query::prepare_rename] prepare_rename: not a function on function definition site: Some(FigureElem { body: Text(Test1), placement: Some(None), caption: Some(None), kind: Some(Custom(Elem(Element(image)))), supplement: Some(Custom(Some(Content(Text(Figure))))), numbering: Some(Some(Pattern(NumberingPattern { pieces: [(\"\", Arabic, Lower)], suffix: \"\", trimmed: false }))), gap: Some(0.65em), outlined: Some(true), counter: Some(Some(Counter(Selector(Elem(Element(figure), Some([(3, Func(image))])))))) })\n"
2024-08-20T12:40:46.486 helix_lsp::transport [ERROR] tinymist err <- "[2024-08-20T10:40:46Z INFO  sync_lsp] handled  textDocument/prepareRename - (5) in 176.06µs\n"
```

Code logs
```
[2024-08-20T10:42:55Z INFO  tinymist_query::prepare_rename] prepare_rename: not a function on function definition site: Some(FigureElem { body: Sequence [SpaceElem, Text(Let), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: LrElem { size: None, body: Sequence [Text((), Text(X), Text(,), SpaceElem, AttachElem { base: #Set(equation.variant: Cal)
    Text(B), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, Text(,), SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, Text())] } }, SpaceElem, Text(and), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: LrElem { size: None, body: Sequence [Text((), Text(Y), Text(,), SpaceElem, AttachElem { base: #Set(equation.variant: Cal)
    Text(B), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, Text(,), SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, Text())] } }, SpaceElem, Text(be two), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: Text(σ) }, Text(-finite measure spaces, and let), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: Sequence [Text(f), Text(:), SpaceElem, Text(X), SpaceElem, Text(×), SpaceElem, Text(Y), SpaceElem, Text(→), SpaceElem, LrElem { size: None, body: Sequence [Text([), Text(0), Text(,), SpaceElem, Text(+), Text(∞), Text(])] }] }, SpaceElem, Text(be measurable with respect to), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: Sequence [AttachElem { base: #Set(equation.variant: Cal)
    Text(B), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, SpaceElem, Text(×), SpaceElem, AttachElem { base: #Set(equation.variant: Cal)
    Text(B), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }] }, Text(. Then), Text(:), SpaceElem, ListItem { body: Sequence [Text(The functions), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: Sequence [Text(x), SpaceElem, Text(↦), SpaceElem, AttachElem { base: Text(∫), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, SpaceElem, Sequence [Text(f), LrElem { size: None, body: Sequence [Text((), Text(x), Text(,), SpaceElem, Text(y), Text())] }], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, LrElem { size: None, body: Sequence [Text((), Text(y), Text())] }] }, SpaceElem, Text(and), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: Sequence [Text(y), SpaceElem, Text(↦), SpaceElem, AttachElem { base: Text(∫), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, SpaceElem, Sequence [Text(f), LrElem { size: None, body: Sequence [Text((), Text(x), Text(,), SpaceElem, Text(y), Text())] }], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, LrElem { size: None, body: Sequence [Text((), Text(x), Text())] }] }, SpaceElem, Text(are well-defined and measurable with respect to), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: AttachElem { base: #Set(equation.variant: Cal)
    Text(B), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None } }, SpaceElem, Text(and), SpaceElem, EquationElem { block: Some(false), numbering: None, number_align: None, supplement: None, body: AttachElem { base: #Set(equation.variant: Cal)
    Text(B), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None } }, Text(, respectively.)] }, SpaceElem, ListItem { body: Sequence [Text(We have), SpaceElem, EquationElem { block: Some(true), numbering: None, number_align: None, supplement: None, body: Sequence [AlignPointElem, SpaceElem, AttachElem { base: Text(∫), t: None, b: Some(Some(Sequence [Text(X), SpaceElem, Text(×), SpaceElem, Text(Y)])), tl: None, bl: None, tr: None, br: None }, SpaceElem, Sequence [Text(f), LrElem { size: None, body: Sequence [Text((), Text(x), Text(,), SpaceElem, Text(y), Text())] }], SpaceElem, LrElem { size: None, body: Sequence [Text((), Sequence [HElem { amount: Rel(0.16666666666666666em), weak: Some(true) }, #Set(equation.italic: Custom(false))
    Text(d)], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, SpaceElem, Text(×), SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, Text())] }, LrElem { size: None, body: Sequence [Text((), Text(x), Text(,), SpaceElem, Text(y), Text())] }, LinebreakElem { justify: None }, SpaceElem, Text(=), SpaceElem, AlignPointElem, SpaceElem, AttachElem { base: Text(∫), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, SpaceElem, LrElem { size: None, body: Sequence [Text((), AttachElem { base: Text(∫), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, SpaceElem, Sequence [Text(f), LrElem { size: None, body: Sequence [Text((), Text(x), Text(,), SpaceElem, Text(y), Text())] }], SpaceElem, Sequence [HElem { amount: Rel(0.16666666666666666em), weak: Some(true) }, #Set(equation.italic: Custom(false))
    Text(d)], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, LrElem { size: None, body: Sequence [Text((), Text(y), Text())] }, Text())] }, SpaceElem, Sequence [HElem { amount: Rel(0.16666666666666666em), weak: Some(true) }, #Set(equation.italic: Custom(false))
    Text(d)], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, LrElem { size: None, body: Sequence [Text((), Text(x), Text())] }, SpaceElem, LinebreakElem { justify: None }, SpaceElem, Text(=), SpaceElem, AlignPointElem, SpaceElem, AttachElem { base: Text(∫), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, SpaceElem, LrElem { size: None, body: Sequence [Text((), AttachElem { base: Text(∫), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, SpaceElem, Sequence [Text(f), LrElem { size: None, body: Sequence [Text((), Text(x), Text(,), SpaceElem, Text(y), Text())] }], SpaceElem, Sequence [HElem { amount: Rel(0.16666666666666666em), weak: Some(true) }, #Set(equation.italic: Custom(false))
    Text(d)], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(X))), tl: None, bl: None, tr: None, br: None }, LrElem { size: None, body: Sequence [Text((), Text(x), Text())] }, Text())] }, SpaceElem, Sequence [HElem { amount: Rel(0.16666666666666666em), weak: Some(true) }, #Set(equation.italic: Custom(false))
    Text(d)], SpaceElem, AttachElem { base: Text(μ), t: None, b: Some(Some(Text(Y))), tl: None, bl: None, tr: None, br: None }, LrElem { size: None, body: Sequence [Text((), Text(y), Text())] }] }, SpaceElem] }, SpaceElem], placement: Some(None), caption: Some(Some(FigureCaption { position: None, separator: None, body: Sequence [Text(Tonelli), SmartQuoteElem { double: Some(false), enabled: None, alternative: None, quotes: None }, Text(s Theorem), SpaceElem, Text((Theorem 1.7.15,), SpaceElem, RefElem { target: Label("tao_2011_IntroductionMeasureTheory"), supplement: None, citation: None, element: None }, Text())], kind: Some(Name("maths")), supplement: Some(Some(Text(Theorem))), numbering: Some(Some(Func(Func(thm-numbering)))), counter: Some(Some(Counter(Selector(Elem(Element(figure), Some([(3, "maths")])))))), figure_location: Some(Some(Location { hash: 167896810197078677029069933691306150431, disambiguator: 0, variant: 0 })) })), kind: Some(Custom(Name("maths"))), supplement: Some(Custom(Some(Content(Text(Theorem))))), numbering: Some(Some(Func(Func(thm-numbering)))), gap: Some(0.65em), outlined: Some(true), counter: Some(Some(Counter(Selector(Elem(Element(figure), Some([(3, "maths")])))))) })
[2024-08-20T10:42:55Z INFO  sync_lsp] handled  textDocument/prepareRename - (147) in 611.22µs
```

--- a/crates/tinymist-analysis/src/syntax/matcher.rs
+++ b/crates/tinymist-analysis/src/syntax/matcher.rs
@@ -701,6 +701,16 @@ impl<'a> SyntaxClass<'a> {
             _ => None,
         }
     }
+
+    pub fn contains_error(&self) -> bool {
+        use SyntaxClass::*;
+        match self {
+            Label { is_error, .. } => *is_error,
+            Normal(kind, _) => *kind == SyntaxKind::Error,
+            Callee(..) | VarAccess(..) => self.node().kind() == SyntaxKind::Error,
+            Ref { .. } | ImportPath(..) | IncludePath(..) => false,
+        }
+    }
 }
 
 /// Classifies node's syntax (inner syntax) that can be operated on by IDE

--- a/crates/tinymist-query/src/fixtures/rename/label.typ
+++ b/crates/tinymist-query/src/fixtures/rename/label.typ
@@ -1,0 +1,7 @@
+/// compile: true
+
+#set heading(numbering: "1.")
+
+= Labeled <title_label>
+
+/* position after */ @title_label

--- a/crates/tinymist-query/src/fixtures/rename/label_indir.typ
+++ b/crates/tinymist-query/src/fixtures/rename/label_indir.typ
@@ -1,0 +1,5 @@
+/// compile: true
+
+#let test1(body) = figure(body)
+#test1([Test1]) <fig:test1>
+/* position after */ @fig:test1

--- a/crates/tinymist-query/src/fixtures/rename/label_indir2.typ
+++ b/crates/tinymist-query/src/fixtures/rename/label_indir2.typ
@@ -1,0 +1,9 @@
+/// compile: true
+
+#let test1(body) = figure(body)
+#test1([Test1]) <fig:test1>
+@fig:test1
+
+#let test2(body) = test1(body)
+#test2([Test2]) <fig:test2>
+/* position after */ @fig:test2

--- a/crates/tinymist-query/src/prepare_rename.rs
+++ b/crates/tinymist-query/src/prepare_rename.rs
@@ -91,8 +91,12 @@ pub(crate) fn prepare_renaming(
             let name = path.get().to_string();
             Some((name, None))
         }
-        // todo: label renaming, bibkey renaming
-        BibEntry(..) | Label(..) | ContentRef(..) => None,
+        // todo: bibkey renaming
+        BibEntry(..) => None,
+        Label(..) | ContentRef(..) => {
+            // todo: label renaming
+            None
+        }
         ImportAlias(..) | Constant(..) | IdentRef(..) | Import(..) | StrName(..) | Spread(..) => {
             None
         }

--- a/crates/tinymist-query/src/prepare_rename.rs
+++ b/crates/tinymist-query/src/prepare_rename.rs
@@ -38,42 +38,53 @@ impl StatefulRequest for PrepareRenameRequest {
         let doc = graph.snap.success_doc.as_ref();
         let source = ctx.source_by_path(&self.path).ok()?;
         let syntax = ctx.classify_for_decl(&source, self.position)?;
-        if matches!(syntax.node().kind(), SyntaxKind::FieldAccess) {
-            // todo: rename field access
-            log::info!("prepare_rename: field access is not a definition site");
+        if bad_syntax(&syntax) {
             return None;
         }
 
         let origin_selection_range = ctx.to_lsp_range(syntax.node().range(), &source);
         let def = ctx.def_of_syntax(&source, doc, syntax.clone())?;
 
-        let (name, range) = prepare_renaming(&syntax, &def)?;
+        let name = prepare_renaming(&syntax, &def)?;
 
         Some(PrepareRenameResponse::RangeWithPlaceholder {
-            range: range.unwrap_or(origin_selection_range),
+            range: origin_selection_range,
             placeholder: name,
         })
     }
 }
 
-pub(crate) fn prepare_renaming(
-    deref_target: &SyntaxClass,
-    def: &Definition,
-) -> Option<(String, Option<LspRange>)> {
-    let name = def.name().clone();
+fn bad_syntax(syntax: &SyntaxClass) -> bool {
+    if matches!(syntax.node().kind(), SyntaxKind::FieldAccess) {
+        // todo: rename field access
+        log::info!("prepare_rename: field access is not a definition site");
+        return true;
+    }
+
+    if syntax.contains_error() {
+        return true;
+    }
+
+    false
+}
+
+pub(crate) fn prepare_renaming(syntax: &SyntaxClass, def: &Definition) -> Option<String> {
+    if bad_syntax(syntax) {
+        return None;
+    }
+
     let def_fid = def.file_id()?;
 
     if WorkspaceResolver::is_package_file(def_fid) {
         crate::log_debug_ct!(
-            "prepare_rename: {name} is in a package {pkg:?}",
+            "prepare_rename: is in a package {pkg:?}, def: {def:?}",
             pkg = def_fid.package(),
         );
         return None;
     }
 
-    let var_rename = || Some((name.to_string(), None));
+    let decl_name = || def.name().clone().to_string();
 
-    crate::log_debug_ct!("prepare_rename: {name}");
     use Decl::*;
     match def.decl.as_ref() {
         // Cannot rename headings or blocks
@@ -82,21 +93,17 @@ pub(crate) fn prepare_renaming(
         // LexicalKind::Mod(Star) => None,
         // Cannot rename expression import
         // LexicalKind::Mod(Module(ModSrc::Expr(..))) => None,
-        Var(..) => var_rename(),
-        Func(..) | Closure(..) => validate_fn_renaming(def).map(|_| (name.to_string(), None)),
+        Var(..) | Label(..) | ContentRef(..) => Some(decl_name()),
+        Func(..) | Closure(..) => validate_fn_renaming(def).map(|_| decl_name()),
         Module(..) | ModuleAlias(..) | PathStem(..) | ImportPath(..) | IncludePath(..)
         | ModuleImport(..) => {
-            let node = deref_target.node().get().clone();
+            let node = syntax.node().get().clone();
             let path = node.cast::<ast::Str>()?;
             let name = path.get().to_string();
-            Some((name, None))
+            Some(name)
         }
         // todo: bibkey renaming
         BibEntry(..) => None,
-        Label(..) | ContentRef(..) => {
-            // todo: label renaming
-            None
-        }
         ImportAlias(..) | Constant(..) | IdentRef(..) | Import(..) | StrName(..) | Spread(..) => {
             None
         }
@@ -139,13 +146,16 @@ mod tests {
         snapshot_testing("rename", &|ctx, path| {
             let source = ctx.source_by_path(&path).unwrap();
 
+            let docs = find_module_level_docs(&source).unwrap_or_default();
+            let properties = get_test_properties(&docs);
+            let graph = compile_doc_for_test(ctx, &properties);
+
             let request = PrepareRenameRequest {
                 path: path.clone(),
                 position: find_test_position(&source),
             };
-            let snap = WorldComputeGraph::from_world(ctx.world.clone());
 
-            let result = request.request(ctx, snap);
+            let result = request.request(ctx, graph);
             assert_snapshot!(JsonRepr::new_redacted(result, &REDACT_LOC));
         });
     }

--- a/crates/tinymist-query/src/rename.rs
+++ b/crates/tinymist-query/src/rename.rs
@@ -1,6 +1,6 @@
 use lsp_types::{
-    DocumentChangeOperation, DocumentChanges, OneOf, OptionalVersionedTextDocumentIdentifier,
-    RenameFile, TextDocumentEdit,
+    AnnotatedTextEdit, ChangeAnnotation, DocumentChangeOperation, DocumentChanges, OneOf,
+    OptionalVersionedTextDocumentIdentifier, RenameFile, TextDocumentEdit,
 };
 use rustc_hash::FxHashSet;
 use tinymist_std::path::{unix_slash, PathClean};
@@ -76,7 +76,7 @@ impl StatefulRequest for RenameRequest {
                 let mut edits: HashMap<Url, Vec<TextEdit>> = HashMap::new();
                 do_rename_file(ctx, def_fid, diff, &mut edits);
 
-                let mut document_changes = edits_to_document_changes(edits);
+                let mut document_changes = edits_to_document_changes(edits, None);
 
                 document_changes.push(lsp_types::DocumentChangeOperation::Op(
                     lsp_types::ResourceOp::Rename(RenameFile {
@@ -94,6 +94,8 @@ impl StatefulRequest for RenameRequest {
                 })
             }
             _ => {
+                let is_label = matches!(def.decl.kind(), DefKind::Reference);
+
                 let references = find_references(ctx, &source, doc, syntax)?;
 
                 let mut edits = HashMap::new();
@@ -108,12 +110,30 @@ impl StatefulRequest for RenameRequest {
                     });
                 }
 
-                log::info!("rename edits: {edits:?}");
+                crate::log_debug_ct!("rename edits: {edits:?}");
 
-                Some(WorkspaceEdit {
-                    changes: Some(edits),
-                    ..Default::default()
-                })
+                if !is_label {
+                    Some(WorkspaceEdit {
+                        changes: Some(edits),
+                        ..Default::default()
+                    })
+                } else {
+                    let change_id = "Typst Rename Labels";
+
+                    let document_changes = edits_to_document_changes(edits, Some(change_id));
+
+                    let change_annotations = Some(create_change_annotation(
+                        change_id,
+                        true,
+                        Some("The language server searched the labels ambiguously".to_string()),
+                    ));
+
+                    Some(WorkspaceEdit {
+                        document_changes: Some(DocumentChanges::Operations(document_changes)),
+                        change_annotations,
+                        ..Default::default()
+                    })
+                }
             }
         }
     }
@@ -300,17 +320,45 @@ impl RenameFileWorker<'_> {
 
 pub(crate) fn edits_to_document_changes(
     edits: HashMap<Url, Vec<TextEdit>>,
+    change_id: Option<&str>,
 ) -> Vec<DocumentChangeOperation> {
     let mut document_changes = vec![];
 
     for (uri, edits) in edits {
         document_changes.push(lsp_types::DocumentChangeOperation::Edit(TextDocumentEdit {
             text_document: OptionalVersionedTextDocumentIdentifier { uri, version: None },
-            edits: edits.into_iter().map(OneOf::Left).collect(),
+            edits: edits
+                .into_iter()
+                .map(|edit| match change_id {
+                    Some(change_id) => OneOf::Right(AnnotatedTextEdit {
+                        text_edit: edit,
+                        annotation_id: change_id.to_owned(),
+                    }),
+                    None => OneOf::Left(edit),
+                })
+                .collect(),
         }));
     }
 
     document_changes
+}
+
+pub(crate) fn create_change_annotation(
+    label: &str,
+    needs_confirmation: bool,
+    description: Option<String>,
+) -> HashMap<String, ChangeAnnotation> {
+    let mut change_annotations = HashMap::new();
+    change_annotations.insert(
+        label.to_owned(),
+        ChangeAnnotation {
+            label: label.to_owned(),
+            needs_confirmation: Some(needs_confirmation),
+            description,
+        },
+    );
+
+    change_annotations
 }
 
 #[cfg(test)]
@@ -323,14 +371,17 @@ mod tests {
         snapshot_testing("rename", &|ctx, path| {
             let source = ctx.source_by_path(&path).unwrap();
 
+            let docs = find_module_level_docs(&source).unwrap_or_default();
+            let properties = get_test_properties(&docs);
+            let graph = compile_doc_for_test(ctx, &properties);
+
             let request = RenameRequest {
                 path: path.clone(),
                 position: find_test_position(&source),
                 new_name: "new_name".to_string(),
             };
-            let snap = WorldComputeGraph::from_world(ctx.world.clone());
 
-            let mut result = request.request(ctx, snap);
+            let mut result = request.request(ctx, graph);
             // sort the edits to make the snapshot stable
             if let Some(r) = result.as_mut().and_then(|r| r.changes.as_mut()) {
                 for edits in r.values_mut() {

--- a/crates/tinymist-query/src/will_rename_files.rs
+++ b/crates/tinymist-query/src/will_rename_files.rs
@@ -1,6 +1,4 @@
-use lsp_types::ChangeAnnotation;
-
-use crate::{do_rename_file, edits_to_document_changes, prelude::*};
+use crate::{create_change_annotation, do_rename_file, edits_to_document_changes, prelude::*};
 
 /// Handle [`workspace/willRenameFiles`] request is sent from the client to the
 /// server.
@@ -37,25 +35,21 @@ impl StatefulRequest for WillRenameFilesRequest {
             })
             .collect::<Option<Vec<()>>>()?;
         log::info!("did rename edits: {edits:?}");
-        let document_changes = edits_to_document_changes(edits);
+        let document_changes = edits_to_document_changes(edits, None);
         if document_changes.is_empty() {
             return None;
         }
 
-        let mut change_annotations = HashMap::new();
-        change_annotations.insert(
-            "Typst Rename Files".to_string(),
-            ChangeAnnotation {
-                label: "Typst Rename Files".to_string(),
-                needs_confirmation: Some(true),
-                description: Some("Rename files should update imports".to_string()),
-            },
-        );
+        let change_annotations = Some(create_change_annotation(
+            "Typst Rename Files",
+            true,
+            Some("Renaming files should update imports".to_string()),
+        ));
 
         Some(WorkspaceEdit {
             changes: None,
             document_changes: Some(lsp_types::DocumentChanges::Operations(document_changes)),
-            change_annotations: Some(change_annotations),
+            change_annotations,
         })
     }
 }

